runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java

```diff
@@ -18,107 +18,230 @@
 
-  private StreamingRuntimeContext runtime = null;
-  private StreamSource.ManualWatermarkContext<WindowedValue<T>> context = null;
+  /**
+   * Keep the options so that we can initialize the readers.
+   */
+  private final SerializedPipelineOptions serializedOptions;
 
-  private volatile boolean isRunning = false;
+  /**
+   * For snapshot and restore.
+   */
+  private final Coder<CheckpointMarkT> checkpointMarkCoder;
 
-  private final SerializedPipelineOptions serializedOptions;
+  /**
+   * The split sources. We split them in the constructor to ensure that all parallel
+   * sources are consistent about the split sources.
+   */
+  private List<? extends UnboundedSource<OutputT, CheckpointMarkT>> splitSources;
 
-  /** Instantiated during runtime **/
-  private transient UnboundedSource.UnboundedReader<T> reader;
+  /**
+   * Make it a field so that we can access it in {@link #trigger(long)} for
+   * emitting watermarks.
+   */
+  private transient List<UnboundedSource.UnboundedReader<OutputT>> readers;
 
-  public UnboundedSourceWrapper(PipelineOptions pipelineOptions, Read.Unbounded<T> transform) {
-    this.name = transform.getName();
-    this.serializedOptions = new SerializedPipelineOptions(pipelineOptions);
-    this.source = transform.getSource();
-  }
+  /**
+   * Initialize here and not in run() to prevent races where we cancel a job before run() is
+   * ever called or run() is called after cancel().
+   */
+  private volatile boolean isRunning = true;
 
-  public String getName() {
-    return this.name;
-  }
+  /**
+   * Make it a field so that we can access it in {@link #trigger(long)} for registering new
+   * triggers.
+   */
+  private transient StreamingRuntimeContext runtimeContext = null;
+
+  /**
+   * Make it a field so that we can access it in {@link #trigger(long)} for emitting
+   * watermarks.
+   */
+  private transient StreamSource.ManualWatermarkContext<WindowedValue<OutputT>> context = null;
+
+  private transient byte[] restoredState = null;
+
+  public UnboundedSourceWrapper(
+      PipelineOptions pipelineOptions,
+      UnboundedSource<OutputT, CheckpointMarkT> source,
+      int parallelism) throws Exception {
+    this.serializedOptions = new SerializedPipelineOptions(pipelineOptions);
 
-  WindowedValue<T> makeWindowedValue(T output, Instant timestamp) {
-    if (timestamp == null) {
-      timestamp = BoundedWindow.TIMESTAMP_MIN_VALUE;
+    if (source.requiresDeduping()) {
+      LOG.warn("Source {} requires deduping but Flink runner doesn't support this yet.", source);
     }
-    return WindowedValue.of(output, timestamp, GlobalWindow.INSTANCE, PaneInfo.NO_FIRING);
+
+    checkpointMarkCoder = source.getCheckpointMarkCoder();
+    // get the splits early. we assume that the generated splits are stable,
+    // this is necessary so that the mapping of state to source is correct
+    // when restoring
+    splitSources = source.generateInitialSplits(parallelism, pipelineOptions);
```

Comment:
Thanks for the long write-up. This is pretty much how it works, yes. Except for some details in when the checkpointing happens and how we restore. (Checkpointing happens by pushing barriers through the topology. When restarting we always restart the whole topology from the last consistent checkpoint.)

I think we won't need a call for now, but thanks for the offer! With the knowledge that the order of the splits is persistent we can move the splitting to the parallel operators at runtime.